### ISSUE TITLE
Surface Scrolling events

### ISSUE THREAD

**Issue by PaulWaldo** [fyne-io/fyne#3442]
### Checklist

- [X] I have searched the [issue tracker](https://github.com/fyne-io/fyne/issues) for open issues that relate to the same feature, before opening a new one.
- [X] This issue only relates to a single feature. I will open new issues for any other features.

### Is your feature request related to a problem?

I have two tables that are related.  When one is scrolled horizontally, I want the other to scroll also.  

## Two tables laid out together
![CleanShot 2022-11-30 at 12 23 17@2x](https://user-images.githubusercontent.com/12105941/204867360-b4f02aa9-f817-41ea-8d49-ad580c3ddc80.png)

## Tables become unsynchronized on a scroll of one

![CleanShot 2022-11-30 at 12 23 33@2x](https://user-images.githubusercontent.com/12105941/204867497-a6f2a480-a4cf-4e05-9eea-753a667794b3.png)

### Is it possible to construct a solution with the existing API?

Not that I can think of.  Scrolling events cannot be caught.

### Describe the solution you'd like to see.

For this particular use-case, binding tables together would be simplest.  If the scoll events were surfaced, I'd be happy to watch for them and manually scroll the other.

**Comment by andydotxyz** [comment]
Hmm, not sure that I agree this is not possible - and adding more APIs to do as you ask will result in:

user scrolls -> driver creates the scroll event -> scroll the table -> table calls out to user scroll event

if you want to make table behave in some way that it controls another you could try extending `Table` and intercept the scroll events which are passed into the table in the first place?

Just a suggestion, not sure if it works for your particular code / setup.

**Comment by ddkwork** [comment]
i think export table's scroll can be fix this problem.

**Comment by andydotxyz** [comment]
Exposing inner workings of a widget like that open it to being broken in unexpected ways. If table hooks into scroll events and you then replace those hooks then table could stop working. This seems bad and is one of the reasons why we avoid exposing inner detail. A widgets API should be based on its functionality and not composed of the items it uses to render.

**Comment by ddkwork** [comment]
> Exposing inner workings of a widget like that open it to being broken in unexpected ways. If table hooks into scroll events and you then replace those hooks then table could stop working. This seems bad and is one of the reasons why we avoid exposing inner detail. A widgets API should be based on its functionality and not composed of the items it uses to render.

Then, can you provide several predefined scroll bar callback functions, such as whether to hide the horizontal and vertical scrolling canvas, which is used for the header, and the callback of horizontal and vertical scrolling of a table body? When the table body rows is horizontally scrolled, the callback header is horizontally scrolled with the same size index size, which realizes the synchronous scrolling of the header and the table body. The horizontal and vertical scrolling canvas of the header should always be hidden. Please correct me if my thinking is wrong.

**Comment by ddkwork** [comment]
Of course, it would be better if a container is provided directly inside the table control to present the header. You only need to copy the layout of the first row of the table body and traverse it to make each column of the header and the table body equal in width or height. If each column of the provided container is a button, you can realize more details, such as sorting each column by pressing the event callback of the button control in each column of the header, calculating the sum of column values, filtering the contents without columns, and so on. The realization of the header will greatly expand the functions of the table. Similar to this, there is the choice of the control type of the table body cell. label is used to present ordinary data, and entry is used to edit and refresh the cell. Can't the onchange callback of entry do this? In addition, cell's single mouse event can also be used to fix the current coordinates of scroll bars, just like Microsoft's excell editor. Finally, the table implementation seems to be short of the behavior of select row, and the features of whether to hide row separators to display more data, such as the decoding display of disassembly engine.

**Comment by andydotxyz** [comment]
> Then, can you provide several predefined scroll bar callback functions, such as whether to hide the horizontal and vertical scrolling canvas, which is used for the header, and the callback of horizontal and vertical scrolling of a table body? When the table body rows is horizontally scrolled, the callback header is horizontally scrolled with the same size index size, which realizes the synchronous scrolling of the header and the table body. The horizontal and vertical scrolling canvas of the header should always be hidden. Please correct me if my thinking is wrong.

This sounds like a bunch of API workaround where it would be better for us to implement headers properly?

**Comment by andydotxyz** [comment]
> Similar to this, there is the choice of the control type of the table body cell. label is used to present ordinary data, and entry is used to edit and refresh the cell. Can't the onchange callback of entry do this? In addition, cell's single mouse event can also be used to fix the current coordinates of scroll bars, just like Microsoft's excell editor.

I don't quite understand this - perhaps you can open a new feature request to expand on it fully.



> Finally, the table implementation seems to be short of the behavior of select row, and the features of whether to hide row separators to display more data, such as the decoding display of disassembly engine.

Is a data table really the right widget for outputting hex codes from dissassembly? I would think monospace labels, Entry or the TextGrid would be much better suited.

**Comment by ddkwork** [comment]
> > Similar to this, there is the choice of the control type of the table body cell. label is used to present ordinary data, and entry is used to edit and refresh the cell. Can't the onchange callback of entry do this? In addition, cell's single mouse event can also be used to fix the current coordinates of scroll bars, just like Microsoft's excell editor.
> 
> I don't quite understand this - perhaps you can open a new feature request to expand on it fully.
> 
> 
> 
> > Finally, the table implementation seems to be short of the behavior of select row, and the features of whether to hide row separators to display more data, such as the decoding display of disassembly engine.
> 
> Is a data table really the right widget for outputting hex codes from dissassembly? I would think monospace labels, Entry or the TextGrid would be much better suited.

Sorry for the late reply. Some time ago, every time I mentioned the problem of the table header, I was inexplicably closed. I think I should study it further for a while. Thank you for this great project.

**Comment by PaulWaldo** [comment]
> This sounds like a bunch of API workaround where it would be better for us to implement headers properly?

@andydotxyz That sounds like the ideal solution, as a number of people have tried to cobble together their own solution.  Headers are integral to the concept of a table, and I think it is in high demand.

I know that is a probably a large job, so I propose an interim solution:

* Add a new function to `Table` that allows programmatic scrolling to an X, Y position.  This is similar to the existing `func (t *Table) ScrollToBottom() `
* Create a `ScrollListener` that yields the new scroll position and allow these listeners to be registered with the table.

I believe these two changes would allow Tables to be synchronized yet prevent too much mucking around with their internals.  What are your thoughts?

**Comment by andydotxyz** [comment]
> Sorry for the late reply. I vaguely remember that some time ago, anyone who mentioned the question of the header of the table would be closed for no reason.

Hopefully this is just a mis-remembering. No issue should every be closed for no reason. Anything that won't be worked on will have a pretty good justification, and probably a discussion as well!

**Comment by andydotxyz** [comment]
> I believe these two changes would allow Tables to be synchronized yet prevent too much mucking around with their internals. What are your thoughts?

I'm not sure that an interim solution is wise - we should focus on desired outcomes rather than enabling workarounds.
If we need to surface events for items that can scroll then, for consistency, this would be required for Table, Tree, List and Entry in all - so it's a fair bit of work but also essential that we get the API right so it works for everything.
rather than event listeners we go more for callbacks, so the equivalent is likely `OnScrolled` rather than `ScrollListener`, mimicking the `Scroll` widget I guess

**Comment by PaulWaldo** [comment]
That would be fine as long as the OnScrolled allowed the callback to determine where it was scrolled and apply a corresponding change to another table

**Comment by PaulWaldo** [comment]
@andydotxyz I tried wrapping both tables together inside a Scrolling container (`NewHScroll`) but that did not seem to have any effect.  I guess the internal scroller for each table overrides the behavior of the container they are in?

**Comment by andydotxyz** [comment]
> I guess the internal scroller for each table overrides the behavior of the container they are in?

In effect yes - each type of interaction works only on the child-most item in the tree. That avoids tapping a button under a button by mistake, for example.
However in this instance a scroller outside the table scroller would only even be possible if you expand the table to display the entire content, completely eliminating the optimisations baked into that widget.
